client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,330 @@
+    entry_filepath: &'static str,
+    table_filepath: &'static str,
+) -> Option<SMBiosData> {
+    match table_load_from_device(entry_filepath, table_filepath) {
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("Failed to load SMBIOS data: {}.", e);
+            None
+        }
+    }
+}
+
+pub fn collect_bios_info(bios_info: &SMBiosInformation) -> Result<devices::Bios> {
+    let release_date_str = bios_info.release_date().to_string();
+    let release_date_format = format_description!("[month]/[day]/[year]");
+    let release_date_parsed = Date::parse(&release_date_str, &release_date_format)?;
+    let output_format = format_description!("[year]-[month]-[day]");
+    let release_date = release_date_parsed.format(&output_format)?;
+
+    let firmware_revision = match (
+        bios_info.system_bios_major_release(),
+        bios_info.system_bios_minor_release(),
+    ) {
+        (Some(major), Some(minor)) => Some(format!("{}.{}", major, minor)),
+        _ => None,
+    };
+
+    let bios = devices::Bios {
+        firmware_revision,
+        release_date: Some(release_date),
+        revision: Some(bios_info.version().to_string()),
+        vendor: bios_info.vendor().to_string(),
+        version: bios_info.version().to_string(),
+    };
+    Ok(bios)
+}
+
+/// Retrieve CPU information from SMBIOS
+pub fn collect_processor_info_smbios(
+    processor_info: &SMBiosProcessorInformation,
+    max_cpu_frequency_filepath: &'static str,
+) -> Result<devices::Processor> {
+    let cpu_id = CpuId::new(processor_info)?;
+    let cpu_freq = CpuFrequency::from_file(max_cpu_frequency_filepath)
+        .unwrap()
+        .mHz();
+
+    let processor = devices::Processor {
+        codename: cpu_id.codename().unwrap_or_else(|| "Unknown".to_string()),
+        frequency: cpu_freq,
+        manufacturer: processor_info.processor_manufacturer().to_string(),
+        version: processor_info.processor_version().to_string(),
+    };
+    Ok(processor)
+}
+
+/// Retrieve CPU information from cpuinfo file
+pub fn retrieve_processor_info_cpuinfo(
+    cpuinfo_filepath: &'static str,
+    max_cpu_frequency_filepath: &'static str,
+) -> Result<devices::Processor> {
+    let cpu_info = CpuInfo::from_file(cpuinfo_filepath)?;
+    let cpu_freq = CpuFrequency::from_file(max_cpu_frequency_filepath)
+        .unwrap()
+        .mHz();
+    let processor = devices::Processor {
+        codename: String::new(),
+        frequency: cpu_freq,
+        manufacturer: cpu_info.cpu_type,
+        version: cpu_info.model,
+    };
+    Ok(processor)
+}
+
+pub fn collect_chassis_info(
+    chassis_info: &SMBiosSystemChassisInformation,
+) -> Result<devices::Chassis> {
+    let chassis_type = chassis_info
+        .chassis_type()
+        .ok_or("Failed to get chassis type")
+        .unwrap()
+        .to_string();
+    let manufacturer = chassis_info.manufacturer().to_string();
+    let sku = chassis_info.sku_number().to_string();
```

Comment:
On the backend side, the SKU term is used (for example, in C3). However, the `SMBiosSystemChassisInformation` struct has only sku_number function, there is no `sku` field or function there. Also, `sku_number` returns a string, not an integer, so I guess it's fine to keep it this way

Reply:
That's a very surprising return type! If `.sku_number()` is returning a `String` (!), then I think we should be able to drop the `.to_string()` as it will cause an extra allocation*

*assuming no optimisation, but then we shouldn't rely on that anyway :)

Reply:
Well, technically it returns `SMBiosString`, that's why we convert it to `String`

Reply:
That's interesting, I just checked, and it looks like the `SMBiosString` type wraps a `Result<String, SMBiosStringError>`, which is very surprising! It also looks like like in case of errors, `.to_string()` will return the error in the string! I think we need to check whether the `SMBiosString` is actually an error before calling `.to_string()` on it! If we want to discard errors, I think something like this may work:
```rust
let sku = chassis_info.sku_number()
    .ok()
    .map(|sku| sku.to_string());

Reply:
I've replaces `to_string()` with `ok()` calls, which return `String` object